### PR TITLE
fix: amb boost performance

### PR DESCRIPTION
**Pull request by sniperadmin**
this tweaks #349

### CODE REVIEW

**Comment by ivelin** [comment]
This is a nice optimization magic, @sniperadmin !

Is there a way to run lighthouse metrics to show the diff in performance before and after this boost?
I remember that the lighthouse action advertised metrics report in the PR timeline, but I don't see any.

It seems the lighthouse CI still runs against ui.ambianic.ai, which is not telling us much about the effect of a particular PR.

As I was looking around, I also noticed that there is a [plugin ](https://github.com/mfrachet/cypress-audit )to embed lighthouse audits within cypress. That can probably add extra robustness to the test suite, since we can cover pages deeper in the app. WDYT?

**Comment by sniperadmin** [comment]
@ivelin 
I think that your proposed plugin can be used with zero config to GH which is cool,..
I like way of robustness
Agree!

**Comment by sniperadmin** [comment]
image
🤣

**Comment by sniperadmin** [comment]
current lighthouse ci does not support localhost, or may be I did something irrelavent i guess!

I will playaround with the proposed plugin and see

**Comment by sniperadmin** [comment]
tried to test the dist folder, but needs to build it somehow

**Comment by ivelin** [comment]
> tried to test the dist folder, but needs to build it somehow

Right. You can take a look at how Cypress e2e tests run in he main CI github action script. I shared that earlier. Instead of dist, it uses "npm run serve", because a web server is needed to serve the pages that need to be tested.

**Comment by sniperadmin** [comment]
@ivelin 
I think i have reached the solution, I ran it locally on my machine, and it works,.. 
Would like to use this [Netlify plugin](https://github.com/marketplace/actions/wait-for-netlify-preview)
it will give a chance for the current lighthouse setup to function on the new generated deploy..

there is a little challenge which is: the `deployed name` that we will use like in this example:
image

this url for example: https://5eb453766c4dcd0006ca3f1c--happy-franklin-69b6d4.netlify.app/
has a generated sequence `5eb453766c4dcd0006ca3f1c` that cannot be guessed..

is it stored in the repo settings as a variable? and how we recall that?

**Comment by ivelin** [comment]
That's great. 

Were you able to also run Lighthouse against a locally deployed instance on localhost:8080 in the github action environment? The way e2e is setup with cypress in the main CI script?

**Comment by sniperadmin** [comment]
yes CI works on my machine! 🔥 🔥 🔥 🔥 
image

**Comment by sniperadmin** [comment]
just one challenge: need to get the deployed version of PR @ netlify

**Comment by sniperadmin** [comment]
for the e2e is setup with cypress in the main CI script,.. I am still testing it
just all the work with current lighthouse that we added to the repo

**Comment by sniperadmin** [comment]
fixed it here it works
image

File: .github/workflows/lighthouse.yml [hunk 1]
```diff
@@ -12,32 +12,35 @@ on:
...
-        echo Add other actions to build,
-        echo test, and deploy your project.
+        npm i
+        npm run build
+
+    - name: Waiting for 200 from the Netlify Preview
+      uses: jakepartusch/wait-for-netlify-action@v1
+      id: waitFor200
+      with:
+        site_name: "happy-franklin-69b6d4"
```
Comment (ivelin) [inline .github/workflows/lighthouse.yml:36]: Great. I see that you are waiting for the netlify preview to deploy before running lighthouse tests.

File: .github/workflows/lighthouse.yml [hunk 2]
```diff
@@ -12,32 +12,35 @@ on:
...
 
     #runs the lighthouse
     - name: Lighthouse CI Action
       uses: treosh/lighthouse-ci-action@v2
       with:
-        # List of URL(s) to analyze
-        urls: 'https://ui.ambianic.ai/'
+        urls: |
+          $DEPLOY_URL
```
Comment (ivelin) [inline .github/workflows/lighthouse.yml:43]: ok, awesome. You are plugging in the dynamic deploy URL from the netlify plugin output.

File: src/components/NavBar.vue [hunk 3]
```diff
@@ -0,0 +1,213 @@
+<template>
```
Comment (ivelin) [inline src/components/NavBar.vue:1]: nice abstraction of the navbar in its own vue component file.

**Comment by sniperadmin** [comment]
@ivelin 
sure! I am so glad with the progress we made so far!
so the last part is lighthouce CI with cypress?

File: .github/workflows/lighthouse.yml (hunk 1 above)
Comment (sniperadmin) [reply .github/workflows/lighthouse.yml:36]: yes, reached it finally 😆

File: .github/workflows/lighthouse.yml (hunk 2 above)
Comment (sniperadmin) [reply .github/workflows/lighthouse.yml:43]: exactly!

File: src/components/NavBar.vue (hunk 3 above)
Comment (sniperadmin) [reply src/components/NavBar.vue:1]: I am looking to optimize and split it more, but in the future

**Comment by sniperadmin** [comment]
I have a feeling that testing both `localhost` and `deployed app` in the same time will have the same results, but the code for each one is in there anyways, I think we can comment out the one we will not use and leave the other...

That's it for this PR if there are no other suggestions

**Comment by ivelin** [comment]
Fantastic work! Going through the code one more time now...

File: .github/workflows/lighthouse.yml [hunk 4]
```diff
@@ -12,32 +12,43 @@ on:
...
+
+    - name: Waiting for 200 from the Netlify Preview
+      uses: jakepartusch/wait-for-netlify-action@v1
+      id: waitFor200
+      with:
+        site_name: "happy-franklin-69b6d4"
 
-    # Runs a set of commands using the runners shell
-    - name: Run a multi-line script
+    # lighthouse localhost test
```
Comment (ivelin) [inline .github/workflows/lighthouse.yml:38]: @sniperadmin This seems to pull and install lighthouse on the local CI environment , but I don't see how it runs tests against a locally deployed ambianic UI instance? Can you tag the lines that start the local UI instance and then point lighthouse to it? I don't see "npm run serve" nor reference to localhost:8080. What am I missing?

File: .github/workflows/lighthouse.yml [hunk 5]
```diff
@@ -12,32 +12,43 @@ on:
...
+    - name: Lighthouse CI
       run: |
-        echo Add other actions to build,
-        echo test, and deploy your project.
+        npm install -g @lhci/cli@0.3.x
+        lhci autorun --config=.github/lighthouse/lighthouserc.json
```
Comment (ivelin) [inline .github/workflows/lighthouse.yml:42]: I see that you are using lhci autorun, which takes care of starting the web server, but I still don't see the required change in package.json nor do I see the optional command line parameter to enable that automatic run.
https://github.com/GoogleChrome/lighthouse-ci/blob/master/docs/cli.md#autorun
From the docs:

Starting a webserver
To allow autorun to automatically start your webserver, add an npm script named serve:lhci.

example package.json excerpt:

  "scripts": {
    "serve:lhci": "NODE_ENV=production npm run server"
  }
Alternatively, you can supply the custom server initalization command from collect as a flag directly to autorun:

lhci autorun --collect.startServerCommand="rails server -e production"

Comment (ivelin) [reply .github/workflows/lighthouse.yml:42]: This PR size is now XXL. Could you please extract just the Lighthouse CI code in a separate PR so we expedite its review and merge. Thank you!

**Comment by ivelin** [comment]
@sniperadmin I spent all of today to extract the minimum unit of value from this PR into a smaller one (#367 ) that we can merge without breaking the build. I realize now that it is no small task to ensure strict LHCI PWA compliance. It gives me a better appreciation for all your work on optimizing performance to meet LHCI assertions and budgets.

File: .github/workflows/lighthouse.yml (hunk 4 above)
Comment (sniperadmin) [reply .github/workflows/lighthouse.yml:38]: this will need some more commits i guess

File: .github/workflows/lighthouse.yml (hunk 5 above)
Comment (sniperadmin) [reply .github/workflows/lighthouse.yml:42]: agreed!

**Comment by sniperadmin** [comment]
@ivelin Wow! such an effort, thanks!
actually it is not the hardest, but requires a lot of effort to check if it works, and why, and is it what we need or not....  etc .. I will check the PR

**Comment by ivelin** [comment]
@sniperadmin just FYI, I am proceeding with merging this branch with master and will try to pull another chunk into a separate smaller PR that we can review and merge.

**Comment by sniperadmin** [comment]
Copy that! I will check this out

**Pull request merged**